File: ui/log.go
package ui

import (
	"fmt"
	"io"
	"os"
	"strings"
	"microscope/config"
	"microscope/core/editor"
	"microscope/core/log_writer"
)

func RunLog(logFilename string) error {
	s, err := log_writer.GetSerializer(config.Load().INITIAL_SERIALIZER_VERSION)
	if err != nil {
		return err
	}

	err = log_writer.Read(logFilename, func(e editor.LogEntry) bool {
		var b []byte
		b, err = s.Marshal(e)
		if err != nil {
			return false
		}
		_, err = fmt.Fprintln(os.Stdout, strings.TrimSpace(string(b)))
		if err != nil {
			return false
		}
		return true
	})
	if err == io.EOF {
		err = nil
	}
	return err
}

File: src/ui/log.rs
use std::io;

use crate::config;
use crate::core::log_writer::{reader, serializer::get_serializer};

pub fn run_log(log_filename: &str) -> io::Result<()> {
    let cfg = config::load();
    let ser = get_serializer(cfg.initial_serializer_version)?;

    reader::read(log_filename, |entry| {
        match ser.marshal(&entry) {
            Ok(b) => {
                let s = String::from_utf8_lossy(&b);
                println!("{}", s.trim());
                true
            }
            Err(_) => false,
        }
    })?;
    Ok(())
}
